### PR TITLE
Improve statistics

### PR DESCRIPTION
**Pull request by LipuFei**
To enable per-community statistics + some cleanup.

DO NOT MERGE YET. Many changes needed to be done for gumby and tribler.

### CODE REVIEW

File: community.py [hunk 1]
```diff
@@ -250,7 +250,19 @@ def __init__(self, dispersy, master, my_member):
         # _pending_tasks contains all pending calls that should be removed when the community is unloaded.
         self._pending_tasks = {}
 
-    def initialize(self):
+        self._statistics = CommunityStatistics(self)
```
Comment (NielsZeilemaker) [inline community.py]: This is a merge bug, can you remove the attach=True, and documentation from the initialize

File: dispersy.py [hunk 2]
```diff
@@ -1730,8 +1730,10 @@ def store_update_forward(self, possibly_messages, store, update, forward):
                 logger.debug("commit user generated message")
                 self._database.commit()
 
-                self._statistics.created_count += my_messages
-                self._statistics.dict_inc(self._statistics.created, messages[0].meta.name, my_messages)
+                for message in messages:
```
Comment (NielsZeilemaker) [inline dispersy.py]: What happened here? This if is already there.

File: statistics.py [hunk 3]
```diff
@@ -226,25 +273,44 @@ class CommunityStatistics(Statistics):
     def __init__(self, community):
         super(CommunityStatistics, self).__init__()
 
+        self._dispersy = community.dispersy
         self._community = community
+        self.database_id = community.database_id
```
Comment (NielsZeilemaker) [inline statistics.py:259]: Could you make these @properties similar to the DispersyStatistics

File: statistics.py [hunk 4]
```diff
@@ -226,25 +273,44 @@ class CommunityStatistics(Statistics):
...
+        self.classification = community.get_classification()
+        self.cid = community.cid
+        self.mid = community.my_member.mid
+        self.hex_cid = community.cid.encode("HEX")
+        self.hex_mid = community.my_member.mid.encode("HEX")
+
+        self.global_time = None
```
Comment (NielsZeilemaker) [inline statistics.py]: The same goes for these

File: statistics.py [hunk 5]
```diff
@@ -254,9 +320,13 @@ def update(self, database=False):
         self.global_time = self._community.global_time
         now = time()
         self.candidates = [(candidate.lan_address, candidate.wan_address, candidate.global_time)
```
Comment (NielsZeilemaker) [inline statistics.py]: Maybe even these

File: statistics.py [hunk 6]
```diff
@@ -254,9 +320,13 @@ def update(self, database=False):
...
-                           for candidate
-                           in self._community.candidates.itervalues() if candidate.get_category(now) in [u'walk', u'stumble', u'intro']]
+            for candidate in self._community.candidates.itervalues()
+                if candidate.get_category(now) in [u'walk', u'stumble', u'intro']]
         if database:
```
Comment (NielsZeilemaker) [inline statistics.py:310]: Only keeping the database stuff in de update method, which whenever its called goes into the database.

File: dispersy.py (hunk 2 above)
Comment (LipuFei) [inline dispersy.py]: This is to update the statistics per community. So we have to iterate over the messages again.

Comment (NielsZeilemaker) [inline dispersy.py]: All messages in this method are from the same community, hence you could simply do a

```
message.community.statistics.increase_msg_count("created", message.meta.name, len(my_messages)
```

File: dispersy.py [hunk 7]
```diff
@@ -1131,8 +1130,8 @@ def check_member_and_global_time(unique, times, message):
                             # from this batch.
                             pass
                         else:
-                            self._statistics.dict_inc(self._statistics.outgoing, u"-sequence-")
-                            self._endpoint.send([message.candidate], [str(packet)])
+                            self._send_packets([message.candidate], [str(packet)],
+                                message.community, "-caused by missing-sequence-")
```
Comment (NielsZeilemaker) [inline dispersy.py]: caused by check_last_sync

File: dispersy.py [hunk 8]
```diff
@@ -1232,8 +1231,8 @@ def check_double_member_and_global_time(unique, times, message):
                         # apparently the sender does not have this message yet
                         if message.distribution.history_size == 1:
                             packet_id, have_packet = tim.values()[0]
-                            self._statistics.dict_inc(self._statistics.outgoing, u"-sequence-")
-                            self._endpoint.send([message.candidate], [have_packet])
+                            self._send_packets([message.candidate], [have_packet],
+                                message.community, "-caused by missing-sequence-")
```
Comment (NielsZeilemaker) [inline dispersy.py]: caused by check_last_sync

File: dispersy.py [hunk 9]
```diff
@@ -990,8 +989,8 @@ def _check_full_sync_distribution_batch(self, messages):
                         if (global_time, packet) < (message.distribution.global_time, message.packet):
                             # we keep PACKET (i.e. the message that we currently have in our database)
                             # reply with the packet to let the peer know
-                            self._statistics.dict_inc(self._statistics.outgoing, '-duplicate-sequence-')
-                            self._endpoint.send([message.candidate], [packet])
+                            self._send_packets([message.candidate], [packet],
+                                message.community, "-caused by duplicate-sequence-")
```
Comment (NielsZeilemaker) [inline dispersy.py]: caused by check_full_sync

File: statistics.py [hunk 10]
```diff
@@ -226,37 +254,71 @@ class CommunityStatistics(Statistics):
...
+    def dispersy_acceptable_global_time_range(self):
+        return self._community.dispersy_acceptable_global_time_range
+
+    @property
+    def global_time(self):
+        return self._community.global_time
+
+    @property
+    def dispersy_enable_candidate_walker(self):
+        return self._community.dispersy_enable_candidate_walker
```
Comment (NielsZeilemaker) [inline statistics.py]: This is a static value, you can move it into the constructor.

File: statistics.py [hunk 11]
```diff
@@ -226,37 +254,71 @@ class CommunityStatistics(Statistics):
...
+    def increase_delay_msg_count(self, category, value=1):
+        self.msg_statistics.increase_delay_count(category, value)
+        self._dispersy.statistics.msg_statistics.increase_delay_count(category, value)
+
+    @property
+    def acceptable_global_time(self):
+        return self._community.acceptable_global_time
+
+    @property
+    def dispersy_acceptable_global_time_range(self):
```
Comment (NielsZeilemaker) [inline statistics.py]: This is a static value, you can move it into the constructor.

File: statistics.py [hunk 12]
```diff
@@ -226,37 +254,71 @@ class CommunityStatistics(Statistics):
...
+    def global_time(self):
+        return self._community.global_time
+
+    @property
+    def dispersy_enable_candidate_walker(self):
+        return self._community.dispersy_enable_candidate_walker
+
+    @property
+    def dispersy_enable_candidate_walker_responses(self):
+        return self._community.dispersy_enable_candidate_walker_responses
```
Comment (NielsZeilemaker) [inline statistics.py]: This is a static value, you can move it into the constructor.

**Pull request merged**